### ISSUE TITLE
ClientAsyncWriter::Write does not return false on CompletionQueue when stream is closed

### ISSUE THREAD

**Issue by Norman0406** [grpc/grpc#14812]
Hello,

### What version of gRPC and what language are you using?

C++ with gRPC 1.10.0
 
### What operating system (Linux, Windows, …) and version?

Windows
 
### What runtime / compiler are you using (e.g. python version or version of gcc)

MSVC 19.13.26129
 
### What did you do?

I am creating a client to server streaming RPC with the following RPC definition:

`rpc ClientStream(stream Request) returns (Response);`

The client streams multiple requests to the server, which then decides on it's own when to close the stream and sends a response back to the client.
 
### What did you expect to see?

I expected the client's completion queue to return false on the Write operation after the server closed the stream. This assumption is based on the documentation of the synchronous counterpart `bool ClientWriter::Write()`. It says: `\return \a true on success, \a false when the stream has been closed.`
 
### What did you see instead?

The completion queue returns true on the write operation and the client just continues to stream requests, although the server closed the stream.
 
### Anything else we should know about your project / environment?

I compared the behaviour with a python client implementation. Here, the python client gets notified that the stream was closed and displays the correct response.

**Comment by vjpai** [comment]
Hi there,
https://github.com/grpc/grpc/blob/589e2cfe098d6e98d1bfe79c72eb1c4d41a72c01/include/grpcpp/impl/codegen/completion_queue.h#L120 has the authoritative documentation on the meaning of `ok`.  `ok==false` on a Write is a promise that the stream is dead, but it should not be assumed that `ok==true` on a Write means that the stream is still alive. This is all asynchronous, so the writing into flow control buffers might have happened, then the server sent the close, then the application got the ok notification based on the fact that the library put the data into flow control. An `ok==true` on write means that the data has the scope to hit the wire but doesn't indicate that the other side got the data or that the wire hasn't been cut. OTOH, an `ok=false` is a strong promise that the stream is still alive.

IMO, the comment on the sync API object is misleading and should be fixed for clarity. "success" should be redefined in a way consistent with the async API explanation of `ok`.

**Comment by Norman0406** [comment]
Dear Vijay,

You say that `ok==false` is a promise that the stream is dead. But that's not the case for a client-side streaming RPC. If the server closes the stream while the client is still writing to it, `false` is not returned on the client-side write operation.

If not by means of `ok==false`, how else can the client know that the server closed the stream? Server-side we have `AsyncNotifyWhenDone`, but there is no such thing on the `ClientContext`.

**Comment by vjpai** [comment]
Right, that's why I was being very careful with my words. If ok is returned as false, then we know that the stream is dead. The logical inverse of this statement does not follow. In other words, if ok is returned as true, then we don't know if the stream is alive or dead. All we know is that the client believed that the stream was alive when the operation buffered its write, but that may have been an arbitrary amount of time before the application called Next or Next returned. (Side note: did we interact on gitter? If so, I just want to reaffirm what I stated there: it's not a good idea for an application to depend on `ok==false` behavior following a `Write` since even `ok==true` doesn't imply that the other side bothered to read the data).

As for when you can be notified for sure that the server closed the stream, the only way of doing so conclusively is at the Finish.

**Comment by vjpai** [comment]
That said, if you have a specific code segment that you believe would make the specific problem more clear, please send it; I'd be glad to look at it.

**Comment by Norman0406** [comment]
Thanks for this explanation. The twist about the meaning of `ok` is now clear to me. However, I believe we are still missing some functionality here. To explain what I mean, I prepared some code snippets (I will use gists to keep it readable). And yes, I was for briefly active on gitter yesterday ;)

All is based on the following assumptions:
- According to the [gRPC doc](https://grpc.io/docs/guides/concepts.html#cancelling-rpcs), streams can be closed by either side at any time
- It doesn't matter what language gRPC clients and servers are written in. They are interoperable and can be exchanged at will.

My example uses the following simple service definition:
```proto
syntax = "proto3";

package test;

message TestRequest
{
    int64 value = 1;
}

message TestResponse
{
    int64 value = 1;
}

service BasicTestService
{
    rpc Accumulate(stream TestRequest) returns (TestResponse);
}
```

This gist shows a simple test case:
https://gist.github.com/Norman0406/8964bc174eb52a04568ceae4fdf46c4c
A server is created. After it received some number of requests, it sends a result to the client and closes the stream. A stub now writes to the stream. However, the stub has no chance to discover that the server is not listening anymore. From the [documentation of the `Write` function](https://github.com/grpc/grpc/blob/master/include/grpcpp/impl/codegen/sync_stream.h#L123), I would assume that I need to check the return value, but it is always true. The gRPC documentation states that 

> A cancellation terminates the RPC immediately so that no further work is done.

In this case, however, the client has to keep on sending without actually having a listening counterpart anymore.

As a matter of fact, this behavior _is_ supported if we look at different programming languages. I used a python stub to demonstrate that it works as expected:
https://gist.github.com/Norman0406/4b3a4aee43e48a91a69f4856d86813fe
The client outputs:
> Server calculated: 15

This is an example where gRPC behavior differs for different programming languages and is not consistent with the concepts of gRPC. 

Of course there is still the possibility there is a feature that I missed, which I need to implement. But unless that's the case, it looks to me that server-side cancellation of a client-side stream is not (fully) supported by the C++ gRPC library, while it is for other programming languages.

**Comment by vjpai** [comment]
TLDR: I've reopened this issue and am considering it a bug for now because it looks like a bad experience.

I think that our implementation is technically within the spec of the API: since there's no guarantee of ordering between the finish operation at the server and anything at the client in the gists, there's no guarantee that the client needs to end in a timely manner as opposed to keeping on running until the client's flow control buffer fills up (which might be huge). However, the behavior is counterintuitive.

The essence of how this happens and why the languages differ is in what actually happens when creating a client-side streaming object (either ClientAsyncWriter or ClientWriter). In C++, this merely creates an operation to collect the message and status from the server, but doesn't actually issue it or wait on it until the Finish operation is called. This is because the C++ API allows the program to optionally just wait for initial metadata rather than waiting for the entire operation to complete; if the user invokes this operation explicitly, the Finish operation doesn't need to wait for initial metadata, but if the user does not invoke this operation, the collection of initial metadata is added to the work to be done at Finish. In contrast, in Python, creating a _stream_unary RPC immediately fires off the operations that receive initial metadata, message, and status from the server since Python doesn't have a mode to just wait for initial metadata.

Let me also add that in C++, this will be an easier behavior to match for sync operations than async operations since the actual core interactions (described as operations above) are not really reported to the user directly, whereas there is a much tighter coupling in the case of the async API between API calls and internal calls.

**Comment by wizardsd** [comment]
It would be nice to have an example of "file transferring via synchronous gRPC stream", which handles packet loss problems by retransmitting fragments. At least gRPC users would understand the non-obvious parts of C ++ gRPC synchronous streaming.

**Comment by vjpai** [comment]
Thanks for the help in narrowing this problem down. Would you mind posting the updated gists for those other sample cases?

**Comment by vjpai** [comment]
FWIW, your comment also indicates that we have a test coverage gap. I'm going to file a different bug about that SendInitialMetadata thing. Hardly anyone uses client-streaming and we seem to have not covered all of the APIs associated with this case in tests.

**Comment by Norman0406** [comment]
Dear Vijay,

Thank you very much for your effort up to this point. I highly appreciate it. I will test the change also with my code as soon as it is merged.

Please have a look at the following gist, which crashes at the location indicated by the `//NOTE` if the server sends initial metadata:
https://gist.github.com/Norman0406/4185aaf320d5aae80eb0322a666b2885

However, my initial question still holds I believe. Say I have a server that does _not_ send initial metadata. How can my client know that the server closed the stream? With your merged pull request, it would be possible to implement it only for servers that send initial metadata, but not for ones that don't. As written above, my test client continued running for over 5 minutes and never got notified.

**Comment by vjpai** [comment]
Thanks also for filling in a lot of details that have made it possible for us to reproduce this issue and better understand it. I'm going to seek some help on this before getting back to you. For performance reasons, we have expanded flow control buffers (and let them dynamically expand as well) in recent releases, so I want to understand the interaction with that properly before answering the issue about notification.

**Comment by vjpai** [comment]
I just had a flow-control discussion with @ncteisen . Indeed our flow control buffers can get quite large on the client-side. If the buffers do fill up, the return tag will get delayed, but that doesn't seem to be happening in your case. Noah will follow up with some further analysis or suggestions.

**Comment by ncteisen** [comment]
To get more details on flow control buffers, could you run with the tracer enabled?
```
export GRPC_VERBOSITY=debug
export GRPC_TRACE=flowctl
```

That should show what is happening to the windows. They are tuned automatically and can get quite big, so I am not surprised that they are not stalling the stream

**Comment by Norman0406** [comment]
I ran the same test as [in this gist](https://gist.github.com/Norman0406/4185aaf320d5aae80eb0322a666b2885), now with the current master. Shortly after the server has sent the result, flow control messages stopped appearing. This is what I got (sorry for the terrible formatting):
> D0709 21:06:13.932000000  5952 dns_resolver.cc:339] Using native dns resolver
I0709 21:06:13.951000000  5952 server_builder.cc:270] Synchronous server. Num CQs: 1, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10000
D0709 21:06:14.071000000  5952 dns_resolver.cc:280] Start resolving.
I0709 21:06:14.077000000 11592 subchannel.cc:613] New connected subchannel at 0190AB18 for subchannel 019116A8
D0709 21:06:14.077000000 11612 flow_control.cc:110] 01916E30[0][svr] | t updt sent | trw:                         65535, ttw:                       4194304, taw:              65535 -> 4194304, srw:                              , slw:                              , saw:
I0709 21:06:14.082000000 11592 frame_settings.cc:220] 01914818[cli] adding 4128769 for initial_window change
D0709 21:06:14.083000000 11592 flow_control.cc:110] 01915590[0][cli] | t updt recv | trw:              65535 -> 4194304, ttw:                       4194304, taw:                         65535, srw:                              , slw:                              , saw:
D0709 21:06:14.084000000 11592 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194304, ttw:                       4194304, taw:                         65535, srw:                       4194304, slw:                       4194304, saw:                       4194304
D0709 21:06:14.084000000 11592 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194304, ttw:                       4194304, taw:              65535 -> 4194304, srw:                              , slw:                              , saw:
I0709 21:06:14.085000000 11612 frame_settings.cc:220] 019160B8[svr] adding 4128769 for initial_window change
D0709 21:06:14.086000000 11612 flow_control.cc:110] 01916E30[0][svr] | t updt recv | trw:              65535 -> 4194304, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.088000000 11612 flow_control.cc:110] 01916E30[0][svr] | t updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.089000000 11612 flow_control.cc:110] 01916E30[1][svr] | s updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                       4194304, slw:                       4194304, saw:                       4194304
D0709 21:06:14.100000000 11612 flow_control.cc:110] 01916E30[0][svr] | t updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.110000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                       4194304, slw:                       4194304, saw:                       4194304
D0709 21:06:14.112000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194304 -> 4194297, ttw:                       4194304, taw:                       4194304, srw:            4194304 -> 4194297, slw:                       4194304, saw:                       4194304
D0709 21:06:14.114000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                       4194304, slw:                       4194304, saw:                       4194304
D0709 21:06:14.117000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194297, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.128000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194304 -> 4194297, srw:                       4194304, slw:                       4194309, saw:                       4194304
D0709 21:06:14.129000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194297, ttw:                       4194304, taw:                       4194304, srw:                       4194297, slw:                       4194304, saw:                       4194304
D0709 21:06:14.130000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194297, srw:                       4194304, slw:                       4194302, saw:                       4194297
D0709 21:06:14.147000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194297 -> 4194290, ttw:                       4194304, taw:                       4194304, srw:            4194297 -> 4194290, slw:                       4194304, saw:                       4194304
D0709 21:06:14.153000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194290, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.164000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194297 -> 4194290, srw:                       4194304, slw:                       4194309, saw:                       4194297
D0709 21:06:14.165000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194290, srw:                       4194304, slw:                       4194302, saw:                       4194290
D0709 21:06:14.164000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194290, ttw:                       4194304, taw:                       4194304, srw:                       4194290, slw:                       4194304, saw:                       4194304
D0709 21:06:14.173000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194290 -> 4194283, ttw:                       4194304, taw:                       4194304, srw:            4194290 -> 4194283, slw:                       4194304, saw:                       4194304
D0709 21:06:14.174000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194283, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.177000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194290 -> 4194283, srw:                       4194304, slw:                       4194309, saw:                       4194290
D0709 21:06:14.178000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194283, srw:                       4194304, slw:                       4194302, saw:                       4194283
D0709 21:06:14.182000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194283, ttw:                       4194304, taw:                       4194304, srw:                       4194283, slw:                       4194304, saw:                       4194304
D0709 21:06:14.187000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194283 -> 4194276, ttw:                       4194304, taw:                       4194304, srw:            4194283 -> 4194276, slw:                       4194304, saw:                       4194304
D0709 21:06:14.189000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194276, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.191000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194283 -> 4194276, srw:                       4194304, slw:                       4194309, saw:                       4194283
D0709 21:06:14.192000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194276, srw:                       4194304, slw:                       4194302, saw:                       4194276
D0709 21:06:14.191000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194276, ttw:                       4194304, taw:                       4194304, srw:                       4194276, slw:                       4194304, saw:                       4194304
D0709 21:06:14.207000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194276 -> 4194269, ttw:                       4194304, taw:                       4194304, srw:            4194276 -> 4194269, slw:                       4194304, saw:                       4194304
D0709 21:06:14.209000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194269, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.215000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194276 -> 4194269, srw:                       4194304, slw:                       4194309, saw:                       4194276
D0709 21:06:14.216000000 11612 flow_control.cc:110] 01916E30[1][svr] | app st recv | trw:                       4194304, ttw:                       4194304, taw:                       4194269, srw:                       4194304, slw:                       4194302, saw:                       4194269
D0709 21:06:14.216000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194269, ttw:                       4194304, taw:                       4194304, srw:                       4194269, slw:                       4194304, saw:                       4194304
D0709 21:06:14.218000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194269 -> 4194262, ttw:                       4194304, taw:                       4194304, srw:            4194269 -> 4194262, slw:                       4194304, saw:                       4194304
D0709 21:06:14.221000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194262, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:14.224000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data recv | trw:                       4194304, ttw:                       4194304, taw:            4194269 -> 4194262, srw:                       4194304, slw:                       4194309, saw:                       4194269
D0709 21:06:14.224000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194262, ttw:                       4194304, taw:                       4194304, srw:                       4194262, slw:                       4194304, saw:                       4194304
D0709 21:06:16.517000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194262 -> 4194255, ttw:                       4194304, taw:                       4194304, srw:            4194262 -> 4194255, slw:                       4194304, saw:                       4194304
D0709 21:06:14.229000000 11620 flow_control.cc:110] 01916E30[1][svr] | s updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194262, srw:                       4194304, slw:                       4194302, saw:                       4194262
D0709 21:06:16.521000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194255, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.529000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194255, ttw:                       4194304, taw:                       4194304, srw:                       4194255, slw:                       4194304, saw:                       4194304
D0709 21:06:16.532000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194255 -> 4194248, ttw:                       4194304, taw:                       4194304, srw:            4194255 -> 4194248, slw:                       4194304, saw:                       4194304
D0709 21:06:16.539000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194248, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.544000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194248, ttw:                       4194304, taw:                       4194304, srw:                       4194248, slw:                       4194304, saw:                       4194304
D0709 21:06:16.545000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194248 -> 4194241, ttw:                       4194304, taw:                       4194304, srw:            4194248 -> 4194241, slw:                       4194304, saw:                       4194304
D0709 21:06:16.547000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194241, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.549000000  5952 flow_control.cc:110] 01915590[1][cli] | s updt sent | trw:                       4194241, ttw:                       4194304, taw:                       4194304, srw:                       4194241, slw:                       4194304, saw:                       4194304
D0709 21:06:16.525000000 11620 flow_control.cc:110] 01916E30[0][svr] | t updt sent | trw:                       4194304, ttw:                       4194304, taw:            4194262 -> 4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.551000000 11620 flow_control.cc:110] 01916E30[1][svr] | s updt sent | trw:                       4194304, ttw:                       4194304, taw:                       4194304, srw:                       4194304, slw:                       4194302, saw:                       4194302
D0709 21:06:16.554000000 11620 flow_control.cc:110] 01916E30[1][svr] |   data sent | trw:            4194304 -> 4194297, ttw:                       4194304, taw:                       4194304, srw:            4194304 -> 4194297, slw:                       4194302, saw:                       4194302
D0709 21:06:16.562000000 11620 flow_control.cc:110] 01916E30[0][svr] | t updt sent | trw:                       4194297, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.563000000 11620 flow_control.cc:110] 01916E30[0][svr] |   data recv | trw:                       4194297, ttw:                       4194304, taw:            4194304 -> 4194297, srw:                              , slw:                              , saw:
D0709 21:06:16.564000000 11620 flow_control.cc:110] 01916E30[0][svr] |   data recv | trw:                       4194297, ttw:                       4194304, taw:            4194297 -> 4194290, srw:                              , slw:                              , saw:
D0709 21:06:16.566000000 11620 flow_control.cc:110] 01916E30[0][svr] |   data recv | trw:                       4194297, ttw:                       4194304, taw:            4194290 -> 4194283, srw:                              , slw:                              , saw:
D0709 21:06:16.550000000  5952 flow_control.cc:110] 01915590[1][cli] |   data sent | trw:            4194241 -> 4194234, ttw:                       4194304, taw:                       4194304, srw:            4194241 -> 4194234, slw:                       4194304, saw:                       4194304
D0709 21:06:16.571000000  5952 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194234, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.577000000 11620 flow_control.cc:110] 01916E30[0][svr] |   data recv | trw:                       4194297, ttw:                       4194304, taw:            4194283 -> 4194276, srw:                              , slw:                              , saw:
D0709 21:06:16.578000000 11592 flow_control.cc:110] 01915590[1][cli] | s updt recv | trw:                       4194234, ttw:                       4194304, taw:                       4194304, srw:            4194234 -> 4194274, slw:                       4194304, saw:                       4194304
D0709 21:06:16.580000000 11592 flow_control.cc:110] 01915590[0][cli] | t updt recv | trw:            4194234 -> 4194276, ttw:                       4194304, taw:                       4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.582000000 11592 flow_control.cc:110] 01915590[1][cli] |   data recv | trw:                       4194276, ttw:                       4194304, taw:            4194304 -> 4194297, srw:                       4194274, slw:                       4194304, saw:                       4194304
D0709 21:06:16.583000000 11592 flow_control.cc:110] 01915590[0][cli] | t updt sent | trw:                       4194276, ttw:                       4194304, taw:            4194297 -> 4194304, srw:                              , slw:                              , saw:
D0709 21:06:16.584000000 11620 flow_control.cc:110] 01916E30[0][svr] | t updt recv | trw:            4194297 -> 4194304, ttw:                       4194304, taw:                       4194276, srw:                              , slw:                              , saw:

**Comment by Norman0406** [comment]
Was this trace helpful, @ncteisen?

**Comment by muxi** [comment]
Let's meet and talk about it Noah. I'll have some time on Thu.

**Comment by srini100** [comment]
@ncteisen, any update on this issue?

**Comment by ncteisen** [comment]
I will find some time to dig into this next week

**Comment by ncteisen** [comment]
I dug into this today. The problem is that the transport is ignoring the fact that the stream is write closed. This was introduced [here](https://github.com/grpc/grpc/commit/764cf04a13958d72db5a22eb4bbb9370e00777f5#diff-ce37d99bf1a5e7e131e56cee8f3f30c6R1383). 

cc @markdroth, can we meet and talk about this tomorrow? The C++ API needs to return false when write is called on a stream that is has been marked as write_closed from the server.

**Comment by ncteisen** [comment]
I dug into this today. The problem is that the transport is ignoring the fact that the stream is write closed. This was introduced [here](https://github.com/grpc/grpc/commit/764cf04a13958d72db5a22eb4bbb9370e00777f5#diff-ce37d99bf1a5e7e131e56cee8f3f30c6R1383). 

cc @markdroth, can we meet and talk about this tomorrow? The C++ API needs to return false when write is called on a stream that is has been marked as write_closed from the server.

**Comment by srini100** [comment]
@ncteisen, any conclusion on this? Should this be assigned to @markdroth to look into?

**Comment by ncteisen** [comment]
Ahh sorry to have dropped this, it slipped my mind. I can still be the contact, but I need input from mark. I will set up a meeting.

**Comment by pleuba** [comment]
Any update on this issue?

**Comment by kalman5** [comment]
I believe this is what is causing me headache on this: https://github.com/grpc/grpc/issues/17236

**Comment by pleuba** [comment]
I have done a wireshark capture of a client stream with latest version (grpc 1.16.0) that is closed by the server after having received 3 messages. The client continues to Write() without any errors. I can seen the 3 messages from the client, the server response and status, a RST_STREAM, then nothing for subsequent Write() from the client.

The RPC is defined like this: 
> rpc statusChanges(stream IssEvent) returns (google.protobuf.Empty);

What is strange is that for server stream or bidirectional stream, we have an error on the Read
() method as soon as the server close the stream.

@ncteisen Any chance to get this fixed for C++. Can I help in any way? I read that this working for other languages, such as java.

**Comment by Norman0406** [comment]
Any updates here?

**Comment by satheesh-subramanian** [comment]
I have been impacted by this issue and would like to see if there is any progress made here

**Comment by yashykt** [comment]
We recently found a way to reproduce this relatively consistently. https://github.com/grpc/grpc/issues/14812 should fix this issue.
Cheers!

**Comment by yashykt** [comment]
The PR describes the issue and fix in detail

**Comment by Norman0406** [comment]
Hi,

Thanks for addressing this issue and merging the fix into master. I ran a test with the latest version 1.24.0 that has #19868 integrated. It seems as if it only works for physical channels, not for in-process channels. This still breaks our test cases. The following minimal test fails when `inProcess = true`, but succeeds for `inProcess=false`. The implemented test service sums up all the values that the client sends until the sum is above 50 and then returns a result.

Proto file:
```proto
syntax = "proto3";
package test;

message GenericTestRequest
{
    int64 value = 1;
}

message GenericTestResponse
{
    int64 value = 1;
}

service GenericTestService
{
    rpc ClientStream(stream GenericTestRequest) returns (GenericTestResponse);
}
```
Source code:
```cpp

namespace grpc_client_streaming_test
{
class ClientStreamingTestService : public ::test::GenericTestService::Service
{
protected:
    ::grpc::Status ClientStream(
        ::grpc::ServerContext *context,
        ::grpc::ServerReader<::test::GenericTestRequest> *reader,
        ::test::GenericTestResponse *response) override
    {
        std::int64_t sum = 0;
        ::test::GenericTestRequest request;
        while (reader->Read(&request) && sum < 50)
        {
            sum += request.value();
        }

        response->set_value(sum);
        return ::grpc::Status(::grpc::StatusCode::OK, "");
    }
};

TEST_CASE("test client streaming server-side cancellation", "[calls][service][client_stream]")
{
    const bool inProcess = true;

    auto service = std::make_unique<ClientStreamingTestService>();

    ::grpc::ServerBuilder builder;
    builder.RegisterService(service.get());

    auto completionQueue = builder.AddCompletionQueue();
    if (!inProcess)
    {
        builder.AddListeningPort("localhost:50000", ::grpc::InsecureServerCredentials());
    }
    auto server = builder.BuildAndStart();

    auto serverThread = std::thread([&server, &completionQueue]() {
        void *ignored_tag;
        bool ignored_ok;
        while (completionQueue->Next(&ignored_tag, &ignored_ok))
        {
        }

        server->Wait();
    });

    std::shared_ptr<::grpc::ChannelInterface> channel;
    if (inProcess)
    {
        channel = server->InProcessChannel(::grpc::ChannelArguments{});
    }
    else
    {
        channel = grpc::CreateCustomChannel(
            "localhost:50000", ::grpc::InsecureChannelCredentials(), ::grpc::ChannelArguments{});
    }

    // use gRPC stub to access
    auto stub = std::make_unique<::test::GenericTestService::Stub>(channel);

    ::test::GenericTestResponse response;
    ::grpc::ClientContext context;
    auto writer = stub->ClientStream(&context, &response);

    ::test::GenericTestRequest request;
    int value = 0;
    do
    {
        request.set_value(value++);
        // HANG: Write() blocks with an inProcess channel after service closed the stream
    } while (writer->Write(request));

    writer->WritesDone();
    auto status = writer->Finish();
    CHECK(response.value() == 55);

    server->Shutdown();
    completionQueue->Shutdown();

    void *ignored_tag;
    bool ignored_ok;
    while (completionQueue->Next(&ignored_tag, &ignored_ok))
    {
    }

    serverThread.join();
}
} // namespace grpc_client_streaming_test
```

**Comment by yashykt** [comment]
Let's track the inproc bug at https://github.com/grpc/grpc/issues/20409
